File: app/data_model/data_model.go
package data_model

// /////////////////////////////
// Data types
// /////////////////////////////
type DataType int

const (
	Bool DataType = iota
	Int
	Float
	String
	Timestamp
)

func (d DataType) String() string {
	switch d {
	case Bool:
		return "Bool"
	case Int:
		return "Int"
	case Float:
		return "Float"
	case String:
		return "String"
	case Timestamp:
		return "Timestamp"
	}
	return "unknown"
}

///////////////////////////////
// Data Model
///////////////////////////////

type DataModel struct {
	Tables map[string]Table
}

type Table struct {
	Name          string
	Fields        map[string]Field
	LinksToSingle map[string]LinkToSingle
}

type Field struct {
	DataType DataType
}

type LinkToSingle struct {
	LinkedTableName string
	ParentFieldName string
	ChildFieldName  string
}

///////////////////////////////
// Data Access
///////////////////////////////

func (dm DataModel) FieldAt(rootName string, path []string) Field {
	currentRoot := dm.Tables[rootName]

	if len(path) == 1 {
		return currentRoot.Fields[path[0]]
	}

	return dm.FieldAt(currentRoot.LinksToSingle[path[0]].LinkedTableName, path[1:])

}

File: marble-backoffice/src/models/DataModel.ts
export interface LinkToSingle {
  linkedTableName: string;
  parentFieldName: string;
  childFieldName: string;
}

export type DataModelType =
  | "Bool"
  | "Int"
  | "Float"
  | "String"
  | "Timestamp"
  | "unknown";

export interface DataModelField {
  fieldId: string;
  dataType: DataModelType;
  nullable: boolean;
  description: string;
}

export interface DataModelTable {
  tableId: string;
  name: string;
  fields: Record<string, DataModelField>;
  linksToSingle: Record<string, LinkToSingle>;
  description: string;
}

export interface DataModel {
  version: string;
  status: string;
  tables: Record<string, DataModelTable>;
}

export interface CreateDataModelTable {
  tableName: string;
  description: string;
}

export interface CreateDataModelField {
  tableId: string;
  fieldName: string;
  description: string;
  dataType: DataModelType;
  nullable: boolean;
  isEnum: boolean;
}

export interface CreateDataModelLink {
  linkName: string;
  parentTableId: string;
  parentFieldId: string;
  childTableId: string;
  childFieldID: string;
}
